### ISSUE TITLE
InvalidJfrFileException

### ISSUE THREAD

**Issue by c42-arthur** [jvm-profiling-tools/async-profiler#398]
I'm using async-profiler v2.0 with AdoptOpenJDK build 11.0.8+10 running inside an Enterprise LInux 7 container (RedHat UBI 7).
```
$ gdb $JAVA_HOME/lib/server/libjvm.so -ex 'info address UseG1GC'
[...]
Reading symbols from /opt/jdk/lib/server/libjvm.so...done.
Symbol "UseG1GC" is at 0x144f558 in a file compiled without debugging.
```
I'm running the container with argument  "--cap-add=SYS_ADMIN" and "--security-opt seccomp=unconfined".

I've set the following kernel parameters.
```
echo 1 | sudo tee 1 /proc/sys/kernel/perf_event_paranoid
echo 2048 | sudo tee /proc/sys/kernel/perf_event_mlock_kb
```
I'm loading async-profiler with the following Java argument.
```
-agentpath:/tmp/async-profiler-2.0-linux-x64/build/libasyncProfiler.so=start,event=cpu,alloc,lock=10ms,file=profile.jfr
```
I see the following when my process starts.
```
Mar 14 18:32:23 ip-10-0-130-253.ec2.internal docker[5005]: Profiling started
```
Then after application startup I stop the container.

When I try opening the flight recording with JDK Mission Control 7.1.2+b04 I get the following.
```
org.openjdk.jmc.flightrecorder.internal.InvalidJfrFileException
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.createChunkInput(FlightRecordingLoader.java:156)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.access$000(FlightRecordingLoader.java:65)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader$1.getNextChunk(FlightRecordingLoader.java:109)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.createChunkLoader(FlightRecordingLoader.java:295)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.readChunks(FlightRecordingLoader.java:218)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.loadStream(FlightRecordingLoader.java:96)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.loadStream(FlightRecordingLoader.java:77)
	at org.openjdk.jmc.flightrecorder.ui.RecordingLoader.doCreateRecording(RecordingLoader.java:173)
	at org.openjdk.jmc.flightrecorder.ui.RecordingLoader.run(RecordingLoader.java:96)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```
I also get the following error with the jfr2flame converter.
```
Exception in thread "main" java.lang.IndexOutOfBoundsException
	at java.base/java.nio.Buffer.checkIndex(Buffer.java:692)
	at java.base/java.nio.DirectByteBuffer.getLong(DirectByteBuffer.java:858)
	at one.jfr.JfrReader.<init>(JfrReader.java:74)
	at jfr2flame.main(jfr2flame.java:88)
```
I don't have this problem when profiling a "Hello World" class.  Unfortunately I'm unable to share the flight recording.

**Comment by apangin** [comment]
What is the size of the output file?
Can you dump the first 68 bytes of the file? (it's just the JFR header)

**Comment by c42-arthur** [comment]
The file is 70.9 MB.
```
$ od -t x1 -N 68 profile.jfr 
0000000    46  4c  52  00  00  02  00  00  00  00  00  00  00  00  00  00
0000020    00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
0000040    16  6c  49  6b  e2  f1  ef  40  00  00  00  00  00  00  00  00
0000060    00  03  f9  1d  39  17  72  94  00  00  00  00  3b  9a  ca  00
0000100    00  00  00  01                                                
0000104
```

**Comment by apangin** [comment]
The header has not been finalized: the fields that should point to constant pool and metadata are empty.
This apparently means that the process was terminated abnormally (e.g. by SIGKILL), and the recording didn't have a chance to close correctly.

Currently the recording is finalized during the VM shutdown sequence. I'll investigate the possibility to write the recording by chunks, so that even in a case of a fatal error JMC would be able to read a part of the recording.

You may also finalize the recording manually at any time by invoking profiler's `stop` command.

**Comment by c42-arthur** [comment]
I was suspicious of something like this.  I did stop the service but it might have timed out on stop and was killed forcefully by systemd.  I got the same result when I used `kill <pid>` though.  Once I used profile.sh to stop the recording I was able to view it with JMC.  I think it would great if the recording could be salvaged in the case of something like OOM killer.  Thanks for your help and hard work on this project!
